### PR TITLE
Framework

### PR DESCRIPTION
**Pull request by kzhu2**
Basic Classes and Framework
Supports variable size sudoku board + Validates the Board

Currently uses asserts (but may and probably can be changed to cerr & exceptions)

### CODE REVIEW

File: main.cpp [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+            return std::make_pair(false, 0 );
+        }
+        
+        //use a vector to check if value has been seen
+        std::vector<bool> seen(vec.size(), false);
+        size_t num_left = grid_size;
+        
+        for(auto val: vec) {
+            if(val < 0) {
+                continue;
```
Comment (Hao2747) [inline main.cpp:81]: do you need num_left -- here as well?

File: main.cpp [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+private:
+    size_t grid_size; //row & col 
+    size_t block_len; // sqrt(row)
+
+    std::vector<GridRow> grid;
+    
+    //checks is the vectors are possible in sudoku
+    //returns : bool - true or false if possible 
+    //   size_t - number of digits missing in row (only valid when bool is true)
+    std::pair<bool, size_t> check_if_vecs_possible(const
```
Comment (Hao2747) [inline main.cpp:68]: Thank you for your great effort. However, I feel like this function needs to frequently be called when doing search and at the end validating. Each time, you load in an entire vector to check if it is possible, which can be sort of expensive. I don't know if you have seen my email about https://github.com/Hao2747/sudoku-online/blob/master/src/solver_basic.cc code. Their idea is we maintain some states, keeping check of possible availabilities for each row/ colomn/block. Each time you want to add one number, you just need to check if that number is used already. How do you think? we can also setup a zoom meeting to discuss about this.

File: main.cpp (hunk 1 above)
Comment (kzhu2) [reply main.cpp:81]: No.  It should be a better variable name, but num_left is the numbers that are currently not seen.

File: main.cpp (hunk 2 above)
Comment (kzhu2) [reply main.cpp:68]: I was thinking about this, but this looks like part of constraint problem. Will probably add it later.

**Pull request merged**